notebooks/writing_a_trace_transform_cpu_offloading.ipynb

```diff
@@ -0,0 +1,868 @@
+    "    meta=lambda t: TensorProxy(\"offloaded_\" + t.name, like=t, device=thunder.core.devices.Device(\"cpu\")),\n",
+    "    fn=offload_to_cpu_impl,\n",
+    ")\n",
+    "\n",
+    "\n",
+    "# Load the tensor to given GPU\n",
+    "def load_to_gpu_impl(t, device):\n",
+    "    return t.to(device, non_blocking=True)\n",
+    "\n",
+    "\n",
+    "load_to_gpu = offload_ex.register_operator(\n",
+    "    \"load_to_gpu\",\n",
+    "    meta=lambda t, device: TensorProxy(like=t, device=thunder.core.devices.Device(device)),\n",
+    "    fn=load_to_gpu_impl,\n",
+    ")\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "First we will have some helper functions to implement our transformation"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def get_symbols_to_first_or_last_used_variables(symbols, first_used=False):\n",
+    "    \"\"\"\n",
+    "    This function processes a sequence of symbols and determines which variables \n",
+    "    are first/last used by each symbol determined based on argument `first_used`.\n",
+    "    It returns a mapping from variables to the symbols where they were first/last used.\n",
+    "\n",
+    "    Args:\n",
+    "        symbols (iterable): An iterable of symbols\n",
+    "        first_used (bool): Whether to return the map of first used variable to symbol mapping if True otherwise return the map for last used.\n",
+    "                           Defaults to False.\n",
+    "\n",
+    "    Returns:\n",
+    "        variable_to_symbol (dict): A dictionary mapping each variable to the  symbol where it is first/last used based on `first_used` argument.\n",
+    "    \"\"\"\n",
+    "    variable_to_symbol = {}\n",
+    "\n",
+    "    def _mark_first_or_last_use(symbol, variable):\n",
+    "        if not variable in variable_to_symbol:\n",
+    "            variable_to_symbol[variable] = symbol\n",
+    "\n",
+    "    iter_symbols = symbols if first_used else reversed(symbols)\n",
+    "    for symbol in iter_symbols:\n",
+    "        # If this function is used in the combined nvfuser+torch executor, there are no symbols but regions.\n",
+    "        # Regions do not have args, kwargs\n",
+    "        if hasattr(symbol, \"inputs\"):\n",
+    "            variables = tuple(symbol.inputs) + tuple(symbol.outputs)\n",
+    "        else:\n",
+    "            variables = (symbol.flat_variableified_proxy_args) + tuple(symbol.flat_variableified_proxy_outs)\n",
+    "        tree_map(lambda x: _mark_first_or_last_use(symbol, x), variables)\n",
+    "\n",
+    "    return variable_to_symbol\n",
+    "\n",
+    "\n",
+    "def get_symbol_to_idx(symbols):\n",
+    "    '''\n",
+    "    This function returns a map from symbol to it's position in the sequence.\n",
+    "    '''\n",
+    "    return {sym: idx for idx, sym in enumerate(symbols)}\n",
+    "\n",
+    "\n",
```

Comment:
Could you add a bit of an explanation for what move_closer_to_consumer does and why it's needed? I'm assuming that this is for moving the load from cpu calls later in the backward trace because otherwise they are tend to get clumped at the start of the backward pass and this reduces the peak memory savings, but this might not be immediately apparent for the average user without prior experience.